src/main/java/com/google/sps/data/Journey.java

```diff
@@ -0,0 +1,180 @@
+   */
+  public Journey(Point start /** , PathFinder pathFinder*/) {
+    this.start = start;
+    // TODO[ak47na]: add PathFinder interface and the classes that implement it: 
+    // GoogleMapsPathFinder used for production and ManhattanDistancePathFinder for testing.
+    // this.pathFinder = pathFinder;
+    waypoints = new HashMap<>();
+    // Hashmap with keys representing Point objects and values representing their index. E.g. when
+    // a new Point is added to the hashmap, its value will be the equal to the number of elements.
+    // The hashmap is preferred to an ArrayList so that Point objects can be removed efficiently.
+    
+    restrictions = new ArrayList<>();
+    // Array of pairs (a, b) meaning that waypoint with value a in waypoints must be visited before
+    // waypoint b.
+    
+    minTime = 0;
+    numberOfWaypoints = 0;
+  }
+
+  public int getNumberOfWaypoints() {
+    return numberOfWaypoints;
+  }
+  public ArrayList<Pair> getRestrictions() {
+    return restrictions;
+  }
+  public HashMap<Point, Integer> getWaypoints() {
+    return waypoints;
+  }
+
+  /** 
+   * Returns the waypoints of the journey as an ArrayList so that they can be accessed by their 
+   * value in the waypoints hashmap, which is the index in the ArrayList.
+   */
+  private ArrayList<Point> getWaypointsList() {
+    ArrayList<Point> waypointsList = new ArrayList<>();
+    waypointsList.addAll(Collections.nCopies(numberOfWaypoints, start));
+    for (Map.Entry<Point, Integer> entry : waypoints.entrySet()) {
+      waypointsList.set(entry.getValue(), entry.getKey());
+    }
+    return waypointsList;
+  }
+
+  /**
+   * Returns the reverse order in which the waypoints must be visited for an optimal journey.
+   */
+  public ArrayList<Point> getOrderedWaypoints() {
+    return orderedWaypoints;
+  }
+
+  /**
+   * Computes for each waypoint i, pointsAfter[i] = bitmask where bits set to 1 represent the
+   * indices of points that must be visited after point i
+   */
+  private ArrayList<Integer> getPointsAfterFromRestrictions() {
+    throw new UnsupportedOperationException("TODO: Implement this method.");
+  }
+
+  /**
+   * Given bestTime matrix, it computes and returns the optimal order in which points should be
+   * visited starting from the last point.
+   */
+  protected void computeOrderFromMatrix(ArrayList<ArrayList<Pair>> bestTime) {
+    throw new UnsupportedOperationException("TODO: Implement this method.");
+  }
+
+  /**
+   * Using the bestTime matrix computed in findOptimalTimePaths(), it returns true if the minimum
+   * time journey fits in the timeslot and false otherwise.
+   */
+   public boolean findJourneyForTimeslot(DeliverySlot deliverySlot) throws ApiException, IOException, InterruptedException {
+    ArrayList<ArrayList<Pair>> bestTime = findOptimalTimePaths();
+    int allWaypointsConfig = (1 << numberOfWaypoints) - 1;
+    int minTime = bestTime.get((1 << numberOfWaypoints)).get(0).first;
```

Comment:
Ok so lets look at the size of `allWaypointsConfig` here:

- 0 waypoints: `0`
- 1 waypoint: `1`
- 2 waypoints: `3`
- 3 waypoints: `7`
- 4 waypoints: `15`
- 5 waypoints: `31`
- 6 waypoints: `63`

Why does `allWaypointsConfig` need to grow in this way? What is the thinking behind the left-shift?

Reply:
`allWaypointsConfig` is a bitmask that has the first `waypoints.size()` bits set to 1. This is necessary for storing every possible state of waypoints without considering order.
 e.g. if `waypoints.size() = = 4 `and `config = 2^0+2^1+2^3 = 11`, 
        then config represents a state that visits waypoints 0, 1 and 3.

Note: The algorithm doesn't loop through all possible permutations of waypoints because when a courier visits a new waypoint, the only information needed is the previously visited waypoint and which waypoints have been visited.